src/backend/distributed/executor/citus_custom_scan.c

```diff
@@ -204,37 +261,109 @@ CitusModifyBeginScan(CustomScanState *node, EState *estate, int eflags)
+		 */
+		RebuildQueryStrings(workerJob->jobQuery, workerJob->taskList);
 
-			if (planningError != NULL)
-			{
-				RaiseDeferredError(planningError, ERROR);
-			}
+		return;
+	}
 
-			workerJob->taskList = taskList;
-			workerJob->partitionKeyValue = ExtractInsertPartitionKeyValue(jobQuery);
-		}
+	/* at this point, we're about to do the shard pruning */
+	Assert(workerJob->deferredPruning);
+	if (jobQuery->commandType == CMD_INSERT)
+	{
+		HandleDeferredShardPruningForInserts(distributedPlan);
+	}
+	else
+	{
+		HandleDeferredShardPruningForFastPathQueries(distributedPlan);
+	}
+}
+
+
+/*
+ * HandleDeferredShardPruningForInserts does the shard pruning for INSERT
+ * queries and rebuilds the query strings.
+ */
+static void
+HandleDeferredShardPruningForInserts(DistributedPlan *distributedPlan)
+{
+	Job *workerJob = distributedPlan->workerJob;
+	Query *jobQuery = workerJob->jobQuery;
+	DeferredErrorMessage *planningError = NULL;
+
+	/* need to perform shard pruning, rebuild the task list from scratch */
+	List *taskList = RouterInsertTaskList(jobQuery, &planningError);
 
-		RebuildQueryStrings(jobQuery, taskList);
+	if (planningError != NULL)
+	{
+		RaiseDeferredError(planningError, ERROR);
 	}
 
-	/* prevent concurrent placement changes */
-	AcquireMetadataLocks(taskList);
+	workerJob->taskList = taskList;
+	workerJob->partitionKeyValue = ExtractInsertPartitionKeyValue(jobQuery);
+
+	RebuildQueryStrings(jobQuery, workerJob->taskList);
+}
+
+
+/*
+ * HandleDeferredShardPruningForFastPathQueries does the shard pruning for
+ * UPDATE/DELETE/SELECT fast path router queries and rebuilds the query strings.
+ */
+static void
+HandleDeferredShardPruningForFastPathQueries(DistributedPlan *distributedPlan)
+{
+	Assert(distributedPlan->fastPathRouterPlan);
+
+	Job *workerJob = distributedPlan->workerJob;
+
+	bool isMultiShardQuery = false;
+	List *shardIntervalList =
+		TargetShardIntervalForFastPathQuery(workerJob->jobQuery,
+											&workerJob->partitionKeyValue,
+											&isMultiShardQuery, NULL);
 
 	/*
-	 * We are taking locks on partitions of partitioned tables. These locks are
-	 * necessary for locking tables that appear in the SELECT part of the query.
+	 * A fast-path router query can only yield multiple shards when the parameter
+	 * cannot be resolved properly, which can be triggered by SQL function.
 	 */
-	LockPartitionsInRelationList(distributedPlan->relationIdList, AccessShareLock);
+	if (isMultiShardQuery)
+	{
+		ereport(ERROR, (errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
+						errmsg("cannot perform distributed planning on this "
+							   "query because parameterized queries for SQL "
+							   "functions referencing distributed tables are "
+							   "not supported"),
+						errhint("Consider using PL/pgSQL functions instead.")));
+	}
 
-	/* modify tasks are always assigned using first-replica policy */
-	workerJob->taskList = FirstReplicaAssignTaskList(taskList);
+	shardIntervalList = list_make1(shardIntervalList);
```

Comment:
This looked strange to me. It seems to me that it would be better if TargetShardIntervalForFastPathQuery already returned this. Looking at it's other usage we do the same there.

Reply:
Probably better to improve that with a separate PR?

Reply:
unless you have to change tons of lines in that function I think it's fine to do it here, given that you also exported it in this PR.

Reply:
done